File: leetcode-go/Q001-Q100/Q034-Search-for-a-Range/Q034-Search-for-a-Range.go
package Q034_Search_for_a_Range

func searchRange(nums []int, target int) []int {
	first := bsearchFirst(nums, target)
	last := bsearchLast(nums, target)
	return []int{first, last}
}

func bsearchFirst(nums []int, target int) int {
	idx := -1
	l, h := 0, len(nums)-1
	for l <= h {
		m := (l + h) / 2
		if nums[m] >= target {
			h = m - 1
		} else {
			l = m + 1
		}
		if nums[m] == target {
			idx = m
		}
	}

	return idx
}

func bsearchLast(nums []int, target int) int {
	idx := -1
	l, h := 0, len(nums)-1
	for l <= h {
		m := (l + h) / 2
		if nums[m] <= target {
			l = m + 1
		} else {
			h = m - 1
		}
		if nums[m] == target {
			idx = m
		}
	}

	return idx
}

File: src/Y2018/linkedin/Q034_Search_for_a_Range.java
package Y2018.linkedin;

public class Q034_Search_for_a_Range {

    public int[] searchRange(int[] nums, int target) {
        int[] result = new int[]{-1, -1};
        if (nums == null || nums.length == 0) {
            return result;
        }
        int l = 0, r = nums.length - 1;
        while (l < r) {
            int m = l + (r - l) / 2;
            if (nums[m] < target) {
                l = m + 1;
            } else {
                r = m;
            }
        }

        if (nums[l] != target) {
            return result;
        }
        result[0] = l;

        r = nums.length - 1;
        while (l < r) {
            int m = l + (r - l) / 2 + 1;
            if (nums[m] > target) {
                r = m - 1;
            } else {
                l = m;
            }
        }
        result[1] = r;
        return result;
    }

}
